### ISSUE TITLE
Mac Fuse Bug - Inspecting files after commit

### ISSUE THREAD

**Issue by sjezewski** [pachyderm/pachyderm#385]
Specifically, the `TestCommitFinishedReadDir` test fails. Part of the problem is that we're returning a size of zero when the commit is open, but this result is getting cached. This is a bug in OS X fuse caching.

We'll have to piggy back commit info on the inspectCommit requests and do the right thing in the fuse layer. Additionally, we needed a way to invalidate the cache.

After discussing w TV we realized we had a few options:

1- Invalidate the right entries at the Right Time
2 - Disable metadata caching via the `attr` callback

The first one unfortunately was impossible (the way the fuse library works you can't invalidate within a callback ... you get semaphore lock).

The second method works. However, we want to extend this further and disable metadata/dir entry caching in all callbacks that serve either type of information. By disabling these caches, we expect to

a) get correct results 
b) not suffer too much perf hits
c) avoid other weird osxfuse bugs

... I'm most excited about that last one ;-)

Also, I want to generate a fuse spec document. We've hit enough edge cases that we should be explicit about the behavior we support/dont. Things that will be included are:

- each type of system call (e.g mkdir, using redirect '>', using append '>>', ls, etc)
- columns for osx / linux
- columns for commit state: open/closed/none
- some way to denote support / failure / undefined / untested / explicitly unsupported

And this doc should be hosted somewhere (probably GH pages) and included in the make / CI process.

**Comment by tv42** [comment]
I feel like there's still confusion in the air. For example, my definition of Right Time is when a commit is finalized, not when a (now-finalized) commit is looked at via the gRPC API. That is, if an open commit is being mutated on host A, and someone else calls `FinishCommit` on host B, either one of two things should happen:

1. all changes to files in that commit attempted on host A start returning errors (e.g. `EROFS`)
2. `FinishCommit` refuses to proceed and returns an error

Anything else will likely just hide bugs, and sounds like a data loss scenario. Disabling caches decreases the sizes of race windows, does not remove them.

So, I would recommend, in that spec you're proposing, to think of not just the local mounted FUSE filesystem and a pretend-worldview where there is instantaneous global state, but how the distributed system as a whole will operate.

**Comment by sjezewski** [comment]
Yes, _definitely_ part of the spec will be considering the consequences for parallel operations. 

And writing it down will force us to be explicit about the workflows we support.

**Comment by sjezewski** [comment]
Some updates.

---
The Spec

Is on the `fuse-spec` branch, and will be merged into the fix for this issue `fix-mac-fuse` before merging to master.

Goals of the spec:

1) Represent our state of the world (what actions we support/dont support/are undefined)

TL;DR -- Expectations are specified within a test, but not quite at the syscall level

There are many future todos to improve the spec (those are listed below). But we needed to start somewhere. To guarantee that what was reported was actually the behavior we produce, I needed to define the actions at the granularity that we specify expectations. This means the actions are within a single test, but based on how the test is written, not necessarily a single syscall. But, our goal should be to write tests at the syscall level of granularity in the future.

2) Be an easy reference

The spec is generated every time the fuse test package runs. The output is checked against the committed version and the tests fail if these don't match. This is to prevent the spec from getting out of sync w tests running on CI.

The actions are separated into 4 broad types of behavior:

- Root Level Directory
- Repo Level Directories
- Data level:
  - Open Commit
  - Closed Commit

And refer to different parts of the mount, and different states of the commit when accessing data level files/directories.

The reports live at `src/server/pfs/fuse/spec/reports` and exist for linux and mac, since there are differences. We should refer to these in the course of debugging any future fuse issues. If we find different views / combinations more helpful, we can update the output format.

[To see the existing spec html docs, look here](https://github.com/pachyderm/pachyderm/tree/fuse-spec/src/server/pfs/fuse/spec/reports)

------
The actual issue

(fuse cache differences on os x)

Remaining plan of attack:

- create a few syscall level tests for new issue here
  - they should fail initially
  - test for write behavior
  - and attr/readdir interface methods
- then add piggyback logic for commit state on inspect
  - test for this new feature
- make tests pass
  - make sure new tests are added to spec
- add cache busting logic in new places (for `Create`, `Mkdir`, `Symlink`, `Link`, `Mknod`) ... (or at least createDir and mkDir)
    - tests for each
    - and make sure its hooked into the spec


------

Future Todos for the Spec:

- Add test harness to run fuse tests on single mount (user mode) and in parallel (as used by PPS)
    - the behavior for some of the actions may be surprising in different cases, and we want to make sure this is well tested
- add 'user tests' spec which allows running commands like `rm /data/foo && echo 'bar' > data/foo` vs `echo 'bar' > data/foo && rm /data/foo` (which should have the same results)
    - we'll find this very useful when fielding user reported issues
- add syscall level tests
    - we should get in the habit of writing tests this way
    - but there is an argument to be made to test both ends --- e.g. the way a 'normal user' would write the code to read a file (an all in one read file helper) AND the syscall behaviors we expect under the hood

**Comment by sjezewski** [comment]
Update - 

This got hairy, so we're opting for as simple fix as needed. We also decided to change some behavior around deletion (in the scraper branch). Once that lands, I'll retest this issue, but I think it will be fixed by those changes.

**Comment by jdoliner** [comment]
Scraper branch has landed and I can confirm that this test passes when run.
It's slightly strange, the test runs fine if I run it alone but it hangs if I do all the tests in 1 run.
Closing this since I think that's a different bug, all of the things in here have been addressed as far as I know.
